src/core/lib/channel/metrics.h

```diff
@@ -78,46 +79,128 @@ class GlobalInstrumentsRegistry {
     // runs or between different versions.
     InstrumentID index;
   };
-  struct GlobalUInt64CounterHandle : public GlobalInstrumentHandle {};
-  struct GlobalDoubleCounterHandle : public GlobalInstrumentHandle {};
-  struct GlobalUInt64HistogramHandle : public GlobalInstrumentHandle {};
-  struct GlobalDoubleHistogramHandle : public GlobalInstrumentHandle {};
-  struct GlobalCallbackInt64GaugeHandle : public GlobalInstrumentHandle {};
-  struct GlobalCallbackDoubleGaugeHandle : public GlobalInstrumentHandle {};
-  using GlobalCallbackHandle = absl::variant<GlobalCallbackInt64GaugeHandle,
-                                             GlobalCallbackDoubleGaugeHandle>;
+
+  template <ValueType V, InstrumentType I, size_t M, size_t N>
+  struct TypedGlobalInstrumentHandle : public GlobalInstrumentHandle {};
+
+  // C++17 has fold expression that may simplify this.
+  template <ValueType V, InstrumentType I, size_t M, size_t N>
+  static constexpr void IsCallbackGaugeHandle(
```

Comment:
Let's not make this part of the public API surface.  Suggest moving it to `GlobalStatsPluginRegistry` (which is where it's actually used) and making it private.

Also, suggest calling this `AssertIsCallbackGaugeHandle()`.

Reply:
Done.